### DISCUSSION TITLE
Cant figure out how to translate the following into Entity and Dao.

### DISCUSSION THREAD

**Discussion by mashoodChishti** [vitusortner/floor#739] [Q&A]
So I'm making an app in which I have to add a fee structure. And it is as follows
![Screenshot_20230228-152712_Fee Manager](https://user-images.githubusercontent.com/66511065/221828853-c4744442-12fd-4a03-a928-c932881431c3.jpg)

Now the id, name and payment mode is easy to add in the entity as int, String, String respectively. But I cant figure out how to translate the dynamic form of setting detail which will include detail and amount. If you can guide me how to do it. It will be a great help.. Thanks!

**Comment by nickolasdeluca** [comment]
I also have a similar issue, where I need to translate an int value into a class which represents the foreign value associated with the int value stored. I tried to use TypeConverters but they don't seem to allow for int? values, which is what I need for a nullable foreign key.

**Reply by dkaera** [reply]
Hmm, that's interesting. This issue certainly needs to be investigated.

**Reply by nickolasdeluca** [reply]
If you'd like the code sample to investigate, this is the main entity, `Produto`, which has a `foreign key` called `grupo` that would be represented by the `integer` value `codigo` of the entity `Grupo`:
```dart
@Entity(foreignKeys: [
  ForeignKey(
    childColumns: ['grupo'],
    parentColumns: ['codigo'],
    entity: Grupo,
  ),
])
class Produto {
  @PrimaryKey()
  int? codigo;
  String? nome;
  Grupo? grupo;

  Produto({
    this.codigo,
    this.nome,
    this.grupo,
  });
}
```
This is the `Grupo` entity:
```dart
@entity
class Grupo {
  @PrimaryKey()
  int codigo;
  String nome;

  Grupo({
    required this.codigo,
    required this.nome,
  });
}
```
This is the TypeConverter I wrote for the conversion:
```dart
class GrupoConverter extends TypeConverter<Grupo?, int> {
  @override
  Future<Grupo?> decode(int databaseValue) async {
    return await db.grupoDAO.findGrupoById(databaseValue);
  }

  @override
  int? encode(Grupo value) {
    return value.codigo;
  }
}
```
And these are the errors thrown by the dart engine:
```dart
'GrupoConverter.decode' ('Future<Grupo?> Function(int)') is not a valid override of 'TypeConverter.decode' ('Grupo? Function(int)').
'GrupoConverter.encode' ('int? Function(Grupo)') is not a valid override of 'TypeConverter.encode' ('int Function(Grupo?)').
```

If you have any idea why or where I got something wrong, let me know, cause I've ran out of ideas.

**Reply by dkaera** [reply]
Thanks for your sample of code.
This is because you are using `Future` for an overridden method that is not supported by the framework. Good point, I wasn't expecting that use case.

**Reply by nickolasdeluca** [reply]
This would be a nice feature to have. I remember working with something similar with Java's Spring framework, where you could even go further and have multiple representations of the same class to use in different scenarios.

**Reply by dkaera** [reply]
Awesome, I don't think it would be too hard to add this option.

**Reply by nickolasdeluca** [reply]
Should I open an issue with this as a suggestion or you plan to use this one?

**Reply by dkaera** [reply]
yes, you can open a issue and link it to this discussion.
